Compilation errors when building on macOS

Continued https://github.com/zerebubuth/openstreetmap-cgimap/pull/454

To build on macOS, you can use the following workflow as a basis:

https://github.com/deevroman/openstreetmap-cgimap/blob/4d8c571bd797418cb16757e69a142b60f1132796/.github/workflows/macos_14.yml

Problems that I found when compiling on my Apple M1
1.

 ```
/Users/runner/work/openstreetmap-cgimap/openstreetmap-cgimap/src/fcgi_request.cpp:161:56: error: invalid operands to binary expression ('int' and 'std::nullptr_t')
        if (strerror_r(errno, err_buf, sizeof err_buf) == nullptr) {
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~
[ 11%] Linking CXX static library ../../libcgimap_libxml++.a
[ 11%] Built target cgimap_libxml++
1 error generated.
```

The function declaration looks like this: `int	 strerror_r(int __errnum, char *__strerrbuf, size_t __buflen);` It is necessary to compare with 0 instead of nullptr.

2. 
```
In file included from /Users/deevroman/Documents/GitHub/openstreetmap-cgimap/contrib/sjparser/library/sjparser/yajl_parser.cpp:24:
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/contrib/sjparser/library/sjparser/yajl_parser.h:26:10: fatal error: 'yajl/yajl_parse.h' file not found
#include <yajl/yajl_parse.h>
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
```

The problem is in the external dependency of sjparser, but there is an assumption that you need to deal with FindYAJL. cmake. I'll try to debug it. As a workaround, I moved the files from `/opt/homebrew/Cellar/yajl/2.1.0/include/yajl` to the library directory and replaced <yajl/yajl_parse.h> to "yajl_parse. h"

3. 
```
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/main.cpp:86:17: error: use of undeclared identifier 'HOST_NAME_MAX'
  char hostname[HOST_NAME_MAX];
                ^
```

On macOS you will have to set this constant manually https://github.com/eddic/fastcgipp/issues/13#issuecomment-271894079

4.
```
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/../include/cgimap/api06/changeset_upload/node.hpp:35:20: error: call to deleted function 'from_chars'
    auto [_, ec] = std::from_chars(lat.data(), lat.data() + lat.size(), _lat);
                   ^~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__charconv/from_chars_integral.h:37:19: note: candidate function has been explicitly deleted
from_chars_result from_chars(const char*, const char*, bool, int = 10) = delete;
                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__charconv/from_chars_integral.h:223:1: note: candidate template ignored: requirement 'is_integral<double>::value' was not satisfied [with _Tp = double]
from_chars(const char* __first, const char* __last, _Tp& __value) {
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__charconv/from_chars_integral.h:229:1: note: candidate function template not viable: requires 4 arguments, but 3 were provided
from_chars(const char* __first, const char* __last, _Tp& __value, int __base) {
^
1 error generated.
``` 

Similar error in the `set_lon` function. This is an Apple Clang issue: https://stackoverflow.com/questions/74955092/using-the-correct-to-chars-overload

5.
```
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/backend/apidb/pgsql_update.cpp:179:12: error: no matching function for call to 'max'
    return std::max(bbox_size_limit, 0L);
           ^~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__algorithm/max.h:40:1: note: candidate template ignored: deduced conflicting types for parameter '_Tp' ('long long' vs. 'long')
max(_LIBCPP_LIFETIMEBOUND const _Tp& __a, _LIBCPP_LIFETIMEBOUND const _Tp& __b)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__algorithm/max.h:51:1: note: candidate template ignored: could not match 'initializer_list<_Tp>' against 'long long'
max(initializer_list<_Tp> __t, _Compare __comp)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__algorithm/max.h:60:1: note: candidate function template not viable: requires single argument '__t', but 2 arguments were provided
max(initializer_list<_Tp> __t)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/__algorithm/max.h:31:1: note: candidate function template not viable: requires 3 arguments, but 2 were provided
max(_LIBCPP_LIFETIMEBOUND const _Tp& __a, _LIBCPP_LIFETIMEBOUND const _Tp& __b, _Compare __comp)
^
1 error generated.
```

Probable fix: 0L -> 0LL

6. 
```
In file included from /Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/request.cpp:15:
In file included from /opt/homebrew/include/fmt/core.h:5:
In file included from /opt/homebrew/include/fmt/format.h:41:
/opt/homebrew/include/fmt/base.h:1643:3: error: static assertion failed due to requirement 'formattable': Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
  static_assert(
  ^
/opt/homebrew/include/fmt/base.h:2018:20: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::context, request::workflow_status, 0>' requested here
  return {{detail::make_arg<NUM_ARGS <= detail::max_packed_args, Context>(
                   ^
/opt/homebrew/include/fmt/format.h:4365:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::context, request::workflow_status, request::workflow_status, 2UL, 0UL, 255ULL, 0>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/request.cpp:95:35: note: in instantiation of function template specialization 'fmt::format<request::workflow_status &, request::workflow_status &>' requested here
    throw std::runtime_error(fmt::format("Can't move backwards in the request workflow from {} to {}.",
                                  ^
1 error generated.
```

I haven’t come up with a fix yet, so I commented out this code

7. 
```
In file included from /Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/handler.cpp:10:
In file included from /Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/../include/cgimap/handler.hpp:16:
In file included from /Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/../include/cgimap/data_update.hpp:15:
/Users/deevroman/Documents/GitHub/openstreetmap-cgimap/src/../include/cgimap/util.hpp:79:33: error: no member named 'join' in namespace 'fmt'
  return fmt::format("{}", fmt::join(ids, ","));
                           ~~~~~^
1 error generated.
```

Error due to missing header `#include <fmt/ranges.h>`

--- a/.github/workflows/macos_14.yml
+++ b/.github/workflows/macos_14.yml
@@ -1,0 +1,25 @@
+name: Build on macOS 14
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: macos-14
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: |
+          brew install boost cryptopp fmt fcgi yajl libmemcached libpqxx postgresql
+
+      - name: build
+        run: |
+            mkdir build && cd build && \
+            CXXFLAGS="-Wall -Wextra -Wpedantic -Wno-unused-parameter" cmake .. -DBUILD_SHARED_LIBS=OFF -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Release && \
+            make -j${nproc}
+
+      - name: Running openstreetmap-cgimap
+        run: |
+           ./build/openstreetmap-cgimap --help"

--- a/.github/workflows/macos_14.yml
+++ b/.github/workflows/macos_14.yml
@@ -18,8 +18,9 @@ jobs:
         run: |
             mkdir build && cd build && \
             CXXFLAGS="-Wall -Wextra -Wpedantic -Wno-unused-parameter" cmake .. -DBUILD_SHARED_LIBS=OFF -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Release && \
-            make -j${nproc}
+            make -j${nproc} && \
+            ctest --output-on-failure -E "db"
 
       - name: Running openstreetmap-cgimap
         run: |
-           ./build/openstreetmap-cgimap --help"
+           ./build/openstreetmap-cgimap --help

--- a/contrib/sjparser/library/CMakeLists.txt
+++ b/contrib/sjparser/library/CMakeLists.txt
@@ -1,7 +1,4 @@
-find_library(YAJL_LIB yajl)
-if(NOT YAJL_LIB)
-  message(FATAL_ERROR "Can't find yajl library")
-endif()
+find_package(YAJL 2 REQUIRED)
 
 add_library(sjparser STATIC
   sjparser/parsing_error.cpp
@@ -42,7 +39,7 @@ set_target_properties(sjparser PROPERTIES OUTPUT_NAME sjparser_static)
 
 setup_compilation_options(sjparser)
 
-target_link_libraries(sjparser PUBLIC ${YAJL_LIB})
+target_link_libraries(sjparser PUBLIC YAJL::YAJL)
 
 target_include_directories(sjparser PUBLIC
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
@@ -60,7 +57,7 @@ if(SJPARSER_BUILD_SHARED_LIBRARY)
 
   setup_compilation_options(sjparser_shared)
 
-  target_link_libraries(sjparser_shared PUBLIC ${YAJL_LIB})
+  target_link_libraries(sjparser_shared PUBLIC YAJL::YAJL)
 
   target_include_directories(sjparser_shared PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>

--- a/include/cgimap/api06/changeset_upload/node.hpp
+++ b/include/cgimap/api06/changeset_upload/node.hpp
@@ -32,6 +32,7 @@ public:
 
     double _lat;
 
+#if !defined(__APPLE__)
     auto [_, ec] = std::from_chars(lat.data(), lat.data() + lat.size(), _lat);
 
     if (ec == std::errc())
@@ -42,12 +43,26 @@ public:
       throw payload_error("Latitude value is too large");
     else
       throw payload_error("Unexpected parsing error");
+
+#else
+
+    try {
+      _lat = std::stod(lat);
+    } catch (std::invalid_argument &e) {
+      throw payload_error("Latitude is not numeric");
+    } catch (std::out_of_range &e) {
+      throw payload_error("Latitude value is too large");
+    }
+    set_lat(_lat);
+
+#endif
   }
 
   void set_lon(const std::string &lon) {
 
     double _lon;
 
+#if !defined(__APPLE__)
     auto [_, ec] = std::from_chars(lon.data(), lon.data() + lon.size(), _lon);
 
     if (ec == std::errc())
@@ -58,6 +73,20 @@ public:
       throw payload_error("Longitude value is too large");
     else
       throw payload_error("Unexpected parsing error");
+
+#else
+
+    try {
+      _lon = std::stod(lon);
+    } catch (std::invalid_argument &e) {
+      throw payload_error("Longitude is not numeric");
+    } catch (std::out_of_range &e) {
+      throw payload_error("Longitude value is too large");
+    }
+
+    set_lon(_lon);
+
+#endif
   }
 
   void set_lat(double lat) {

--- a/include/cgimap/backend/apidb/pqxx_string_traits.hpp
+++ b/include/cgimap/backend/apidb/pqxx_string_traits.hpp
@@ -127,7 +127,6 @@ PQXX_ARRAY_STRING_TRAITS(std::vector<tile_id_t>);
 PQXX_ARRAY_STRING_TRAITS(std::vector<osm_changeset_id_t>);
 PQXX_ARRAY_STRING_TRAITS(std::set<osm_changeset_id_t>);
 PQXX_ARRAY_STRING_TRAITS(std::vector<std::string>);
-PQXX_ARRAY_STRING_TRAITS(std::vector<bool>);
 
 } // namespace pqxx
 

--- a/include/cgimap/backend/apidb/pqxx_string_traits.hpp
+++ b/include/cgimap/backend/apidb/pqxx_string_traits.hpp
@@ -116,7 +116,7 @@ public:
   template<>                                                            \
   struct string_traits<type> : cgimap::array_string_traits<type> {};    \
                                                                         \
-  template<> std::string const type_name<type>{#type};                  \
+  template<> inline std::string const type_name<type>{#type};           \
 
 
 #endif

--- a/include/cgimap/util.hpp
+++ b/include/cgimap/util.hpp
@@ -27,6 +27,7 @@
 
 #include <fmt/core.h>
 #include <fmt/format.h>
+#include <fmt/ranges.h>
 
 
 inline size_t unicode_strlen(const std::string & s)

--- a/include/cgimap/util.hpp
+++ b/include/cgimap/util.hpp
@@ -30,6 +30,30 @@
 #include <fmt/ranges.h>
 
 
+#if __APPLE__
+// NOTE: <codecvt> is deprecated in C++17 and removed in C++26 (see P2871R3).
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated"
+
+#include <codecvt>
+#include <iomanip>
+#include <stdexcept>
+
+inline std::size_t unicode_strlen(const std::string& utf8)
+{
+
+  try {
+    return std::wstring_convert< std::codecvt_utf8<char32_t>, char32_t >{}.from_bytes(utf8).size();
+  } catch(std::range_error) {
+    throw http::bad_request("Invalid UTF-8 string encountered");
+  }
+}
+
+#pragma clang diagnostic pop
+
+#else
+
 inline size_t unicode_strlen(const std::string & s)
 {
    const char* mbstr = s.c_str();
@@ -43,6 +67,8 @@ inline size_t unicode_strlen(const std::string & s)
 
    return len;
 }
+
+#endif
 
 inline std::string escape(std::string_view input) {
 

--- a/src/backend/apidb/changeset_upload/relation_updater.cpp
+++ b/src/backend/apidb/changeset_upload/relation_updater.cpp
@@ -1239,17 +1239,16 @@ void ApiDB_Relation_Updater::update_current_relations(
 
   m.prepare("update_current_relations",
             R"(   
-        WITH u(id, changeset_id, visible, version) AS (
+        WITH u(id, changeset_id, version) AS (
                 SELECT * FROM
                 UNNEST( CAST($1 AS bigint[]),
                         CAST($2 AS bigint[]),
-                        CAST($3 AS boolean[]),
-                        CAST($4 AS bigint[])
+                        CAST($3 AS bigint[])
                       )
         )
         UPDATE current_relations AS r
         SET changeset_id = u.changeset_id,
-            visible = u.visible,
+            visible = CAST($4 as boolean),
             timestamp = (now() at time zone 'utc'),
             version = u.version + 1
         FROM u
@@ -1260,20 +1259,18 @@ void ApiDB_Relation_Updater::update_current_relations(
 
   std::vector<osm_nwr_signed_id_t> ids;
   std::vector<osm_changeset_id_t> cs;
-  std::vector<bool> visibles;
   std::vector<osm_version_t> versions;
   std::map<osm_nwr_id_t, osm_nwr_signed_id_t> id_to_old_id;
 
   for (const auto &relation : relations) {
     ids.emplace_back(relation.id);
     cs.emplace_back(relation.changeset_id);
-    visibles.push_back(visible);
     versions.emplace_back(relation.version);
     id_to_old_id[relation.id] = relation.old_id;
   }
 
   pqxx::result r =
-      m.exec_prepared("update_current_relations", ids, cs, visibles, versions);
+      m.exec_prepared("update_current_relations", ids, cs, versions, visible);
 
   if (r.affected_rows() != relations.size())
     throw http::server_error("Could not update all current relations");

--- a/src/backend/apidb/changeset_upload/way_updater.cpp
+++ b/src/backend/apidb/changeset_upload/way_updater.cpp
@@ -642,17 +642,16 @@ void ApiDB_Way_Updater::update_current_ways(const std::vector<way_t> &ways,
     return;
 
   m.prepare("update_current_ways", R"(
-      WITH u(id, changeset_id, visible, version) AS (
+      WITH u(id, changeset_id, version) AS (
          SELECT * FROM
          UNNEST( CAST($1 AS bigint[]),
                  CAST($2 AS bigint[]),
-                 CAST($3 AS boolean[]),
-                 CAST($4 AS bigint[])
+                 CAST($3 AS bigint[])
               )
       )
       UPDATE current_ways AS w
       SET changeset_id = u.changeset_id,
-         visible = u.visible,
+         visible = CAST($4 as boolean),
          timestamp = (now() at time zone 'utc'),
          version = u.version + 1
          FROM u
@@ -663,20 +662,18 @@ void ApiDB_Way_Updater::update_current_ways(const std::vector<way_t> &ways,
 
   std::vector<osm_nwr_signed_id_t> ids;
   std::vector<osm_changeset_id_t> cs;
-  std::vector<bool> visibles;
   std::vector<osm_version_t> versions;
   std::map<osm_nwr_id_t, osm_nwr_signed_id_t> id_to_old_id;
 
   for (const auto &way : ways) {
     ids.emplace_back(way.id);
     cs.emplace_back(way.changeset_id);
-    visibles.push_back(visible);
     versions.emplace_back(way.version);
     id_to_old_id[way.id] = way.old_id;
   }
 
   pqxx::result r =
-      m.exec_prepared("update_current_ways", ids, cs, visibles, versions);
+      m.exec_prepared("update_current_ways", ids, cs, versions, visible);
 
   if (r.affected_rows() != ways.size())
     throw http::server_error("Could not update all current ways");

--- a/src/backend/apidb/pgsql_update.cpp
+++ b/src/backend/apidb/pgsql_update.cpp
@@ -176,7 +176,7 @@ uint64_t pgsql_update::get_bbox_size_limit(osm_user_id_t uid)
     auto row = res[0];
     auto bbox_size_limit = row[0].as<int64_t>();
 
-    return std::max(bbox_size_limit, 0L);
+    return (bbox_size_limit < 0 ? 0 : bbox_size_limit);
   }
 }
 

--- a/src/fcgi_request.cpp
+++ b/src/fcgi_request.cpp
@@ -158,7 +158,7 @@ int fcgi_request::accept_r() {
       } else {
         char err_buf[1024];
         out << "error accepting request: ";
-        if (strerror_r(errno, err_buf, sizeof err_buf) == nullptr) {
+        if (strerror_r(errno, err_buf, sizeof err_buf) == 0) {   // NOLINT(modernize-use-nullptr)
           out << err_buf;
         } else {
           out << "error encountered while getting error message";

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -77,6 +77,12 @@ void reload(int) {
   reload_requested = true;
 }
 
+#if __APPLE__
+  #ifndef HOST_NAME_MAX
+    #define HOST_NAME_MAX 255
+  #endif
+#endif
+
 /**
  * make a string to be used as the generator header
  * attribute of output files. includes some instance

--- a/src/request.cpp
+++ b/src/request.cpp
@@ -92,8 +92,8 @@ void request::check_workflow(workflow_status this_stage) {
   } else if (m_workflow_status > this_stage) {
     // oops - workflow is more advanced than the function which called
     // this, so a workflow violation has occurred.
-    throw std::runtime_error(fmt::format("Can't move backwards in the request workflow from {} to {}.",
-                              m_workflow_status, this_stage));
+    throw std::runtime_error(fmt::format("Can't move backwards in the request workflow from {:d} to {:d}.",
+        static_cast<int>(m_workflow_status), static_cast<int>(this_stage)));
   }
 }
 

